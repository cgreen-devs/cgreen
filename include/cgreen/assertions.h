#ifndef ASSERTIONS_HEADER
#define ASSERTIONS_HEADER

#include "internal/assertions_internal.h"
#include "internal/stringify_token.h"

#include <cgreen/constraint.h>
#include <cgreen/filename.h>
#include <cgreen/reporter.h>
#include <stdint.h>

#ifndef __cplusplus
#include <stdbool.h>
#endif

#ifdef __cplusplus
#include <cgreen/cpp_assertions.h>

namespace cgreen {
    extern "C" {
#endif

/*
  Modern style asserts using constraints:

   assert_that(actual, <constraint>(expected));
   assert_that(<expression>);

*/
#define assert_that(...) assert_that_NARG(__VA_ARGS__)(__VA_ARGS__)
#define assert_that_double(actual, constraint) assert_that_double_(FILENAME, __LINE__, STRINGIFY_TOKEN(actual), (double)(actual), (constraint))

#define pass_test() assert_true(true)
#define fail_test(...) assert_true_with_message(false, __VA_ARGS__)


/* Utility: */
int get_significant_figures(void);
void significant_figures_for_assert_double_are(int figures);
void double_relative_tolerance_is(double tolerance);
double get_double_relative_tolerance(void);
void double_absolute_tolerance_is(double tolerance);
double get_double_absolute_tolerance(void);

#include <cgreen/legacy.h>

#ifdef __cplusplus
    }
}
#endif

#endif
