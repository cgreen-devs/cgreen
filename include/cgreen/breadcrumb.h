#ifndef BREADCRUMB_HEADER
#define BREADCRUMB_HEADER

struct CgreenBreadcrumb_ {
    const char **trail;
    int depth;
    int space;
};

typedef struct CgreenBreadcrumb_ CgreenBreadcrumb;

#ifdef __cplusplus
namespace cgreen {
    extern "C" {
#endif

CgreenBreadcrumb *create_breadcrumb(void);
void destroy_breadcrumb(CgreenBreadcrumb *breadcrumb);
void push_breadcrumb(CgreenBreadcrumb *breadcrumb, const char *name);
void pop_breadcrumb(CgreenBreadcrumb *breadcrumb);
const char *get_current_from_breadcrumb(CgreenBreadcrumb *breadcrumb);
int get_breadcrumb_depth(CgreenBreadcrumb *breadcrumb);
void walk_breadcrumb(CgreenBreadcrumb *breadcrumb, void (*walker)(const char *, void *), void *memo);

#ifdef __cplusplus
    }
}
#endif

#endif
