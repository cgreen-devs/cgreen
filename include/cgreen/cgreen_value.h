#ifndef CGREEN_VALUE_HEADER
#define CGREEN_VALUE_HEADER

#include <stdint.h>
#include <stddef.h>

typedef enum {
              CGREEN_INTEGER,
              CGREEN_STRING,
              CGREEN_DOUBLE,
              CGREEN_POINTER,
              CGREEN_BYVALUE,
              CGREEN_BYREFERENCE
} CgreenValueType;

typedef struct {
    CgreenValueType type;
    union {
        intptr_t integer_value;
        double double_value;
        void *pointer_value;
        const char *string_value;
    } value;
    size_t value_size;
} CgreenValue;

#endif
