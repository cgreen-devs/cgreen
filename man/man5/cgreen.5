.mso www.tmac
.TH CGREEN 5
.SH NAME
cgreen \- a framework for C and C++ unit tests and mocking
.SH DESCRIPTION
.B Cgreen
is a framework for creating and running compact and easy-to-read
unittests for C and/or C++.

Here are some of its features:

.IP \[bu] 2
Fluent API resulting in very readable tests
.IP \[bu]
Expressive and clear output using the default reporter
.IP \[bu]
Fully functional mocks, both strict and loose
.IP \[bu]
Each test runs in its own process for test suite robustness
.IP \[bu]
Automatic discovery and running of tests using dynamic library inspection
.IP \[bu]
Extensive and expressive constraints for many datatypes
.IP \[bu]
BDD-flavoured test declarations with Before and After declarations
.IP \[bu]
Extensible reporting mechanism
.IP \[bu]
Fully composable test suites
.IP \[bu]
An isolated test can be run in a single process for debugging


.SH "SEE ALSO"
cgreen-runner(1)
.PP
The full documentation for
.B cgreen-runner
and
.B the Cgreen framework
is in the
.B Cgreen
manual available at
.URL https://github.com/cgreen-devs/cgreen GitHub .
