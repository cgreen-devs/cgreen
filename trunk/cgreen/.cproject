<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<?fileVersion 5.0.0?>

<cproject storage_type_id="org.eclipse.cdt.core.XmlProjectDescriptionStorage2">
	<storageModule moduleId="org.eclipse.cdt.core.settings">
		<cconfiguration id="cdt.managedbuild.toolchain.gnu.base.1521338569">
			<storageModule externalCElementFile="cdt.managedbuild.toolchain.gnu.base.1521338569_org.eclipse.cdt.core.settings" id="cdt.managedbuild.toolchain.gnu.base.1521338569" name="Default"/>
			<storageModule externalCElementFile="cdt.managedbuild.toolchain.gnu.base.1521338569_cdtBuildSystem" version="4.0.0"/>
			<storageModule externalCElementFile="cdt.managedbuild.toolchain.gnu.base.1521338569_org.eclipse.cdt.core.externalSettings"/>
		</cconfiguration>
	</storageModule>
	<storageModule moduleId="cdtBuildSystem" version="4.0.0">
		<project id="cgreen.null.1958735211" name="cgreen"/>
	</storageModule>
	<storageModule moduleId="org.eclipse.cdt.core.LanguageSettingsProviders"/>
	<storageModule moduleId="org.eclipse.cdt.make.core.buildtargets">
		<buildTargets>
			<target name="all" path="" targetID="org.eclipse.cdt.build.MakeTargetBuilder">
				<buildCommand>make</buildCommand>
				<buildArguments/>
				<buildTarget>all</buildTarget>
				<stopOnError>true</stopOnError>
				<useDefaultCommand>false</useDefaultCommand>
				<runAllBuilders>true</runAllBuilders>
			</target>
			<target name="clean" path="" targetID="org.eclipse.cdt.build.MakeTargetBuilder">
				<buildCommand>rm</buildCommand>
				<buildArguments>-rf build</buildArguments>
				<buildTarget/>
				<stopOnError>true</stopOnError>
				<useDefaultCommand>false</useDefaultCommand>
				<runAllBuilders>true</runAllBuilders>
			</target>
		</buildTargets>
	</storageModule>
	<storageModule moduleId="scannerConfiguration">
		<autodiscovery enabled="true" problemReportingEnabled="true" selectedProfileId=""/>
		<scannerConfigBuildInfo instanceId="cdt.managedbuild.toolchain.gnu.base.1521338569;cdt.managedbuild.toolchain.gnu.base.1521338569.1174655132;cdt.managedbuild.tool.gnu.c.compiler.base.349579178;cdt.managedbuild.tool.gnu.c.compiler.input.1942191604">
			<autodiscovery enabled="false" problemReportingEnabled="true" selectedProfileId="org.eclipse.cdt.make.core.GCCStandardMakePerFileProfile"/>
		</scannerConfigBuildInfo>
		<scannerConfigBuildInfo instanceId="cdt.managedbuild.toolchain.gnu.base.1521338569;cdt.managedbuild.toolchain.gnu.base.1521338569.1174655132;cdt.managedbuild.tool.gnu.cpp.compiler.base.1611546606;cdt.managedbuild.tool.gnu.cpp.compiler.input.560078280">
			<autodiscovery enabled="true" problemReportingEnabled="true" selectedProfileId=""/>
		</scannerConfigBuildInfo>
	</storageModule>
</cproject>
