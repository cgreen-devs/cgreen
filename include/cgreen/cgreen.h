#include <stdlib.h>
#include <cgreen/unit.h>
#include <cgreen/suite.h>
#include <cgreen/text_reporter.h>
#include <cgreen/cdash_reporter.h>
#include <cgreen/cute_reporter.h>
#include <cgreen/assertions.h>
#include <cgreen/constraint_syntax_helpers.h>
#include <cgreen/runner.h>
#include <cgreen/boxed_double.h>
