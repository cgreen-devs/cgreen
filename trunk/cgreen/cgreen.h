#include "unit.h"
#include "text_reporter.h"
#include "assertions.h"
#include <stdlib.h>
