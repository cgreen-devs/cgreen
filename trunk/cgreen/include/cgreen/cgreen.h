#include <cgreen/unit.h>
#include <cgreen/text_reporter.h>
#include <cgreen/cute_reporter.h>
#include <cgreen/cdash_reporter.h>
#include <cgreen/assertions.h>
#include <stdlib.h>
