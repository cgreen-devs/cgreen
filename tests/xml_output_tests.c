#include <cgreen/cgreen.h>

Ensure(failing_test_is_listed_by_xml_reporter) {
    assert_that(false);
}

Ensure(passing_test_is_listed_by_xml_reporter) {
    assert_that(true);
}
