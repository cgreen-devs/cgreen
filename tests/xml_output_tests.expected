<?xml version="1.0" encoding="ISO-8859-1" ?>
<testsuite name="xml_output_tests">
</testsuite>
<?xml version="1.0" encoding="ISO-8859-1" ?>
	<testsuite name="xml_output_tests-default">
		<testcase classname="xml_output_tests/default" name="failing_test_is_listed_by_xml_reporter" time="0.00000">
			<failure message="Expected [0] to [be true]">
				<location file="xml_output_tests.c" line="0"/>
			</failure>
		</testcase>
		<testcase classname="xml_output_tests/default" name="passing_test_is_listed_by_xml_reporter" time="0.00000">
		</testcase>
	</testsuite>
